hw/nxagent/Screen.c: Cover Xinerama bounding box corner cases.

sunweaver opened the pull request:
If the agent window is moved around on screen, it can happen
  that it is moved into an invisible area of the real Xserver,
  we calls this "beyond the bounding box".
  .
  If the agent window is partially beyond the bounding box, we
  don't want Xinerama to re-adjust the RandR parameters inside the
  agent. Near the bounding box, the session shall stay intact.
  .
  This means, desktop env wise, the desktop session control
  elements can be moved (with the agent window) into the invisible
  areas of the real Xserver and moved out again without RandR
  events arriving inside the agent session.

Fixes ArcticaProject/nx-libs#662.

sunweaver commented:
@uli42: Now it (see #662) works like charme, but the diff is not funny. And not a joke, either.

@Ionic: I'd dearly suggest you try not to understand the code but rather test the outcome on a Xinerama setup with at least 2 outputs, better 4.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3674,24 +3674,292 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
+    }
+
+    /*
+     * ###### The Y-Coordinate ######
```
I have not checked all the details yet but from the quick glance I took this _should_ be basically the identical code to the X coordinate. So factor that out to a separate function.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -4179,6 +4449,7 @@ int nxagentAdjustRandRXinerama(ScreenPtr pScreen)
 	     refcnt by 1. We decrease it again by calling only
 	     RRModeDestroy() and forget about prevmode */
 	  RRModeDestroy(mymode);
+	  RRCrtcSet(pScrPriv->crtcs[i], mymode, new_x, new_y, RR_Rotate_0, 1, &(pScrPriv->outputs[i]));
```
this is the line I suggested to add, is this still required?

When I started all this I hade code cascades like this, too. But in the end it all lead to basically the same behaviour. So I _feel_ your change being too big. From my POV it is simply a matter of the position not changing in case of a resize. That probably is the only thing that needs to be reconsidered. But I may be totally wrong, I have to dig deeper...

sunweaver commented:
Hi,

On  Fr 02 Mär 2018 15:31:19 CET, Ulrich Sibiller wrote:

> uli42 commented on this pull request.
>
>
>
>> @@ -4179,6 +4449,7 @@ int nxagentAdjustRandRXinerama(ScreenPtr pScreen)
>  	     refcnt by 1. We decrease it again by calling only
>  	     RRModeDestroy() and forget about prevmode */
>  	  RRModeDestroy(mymode);
> +	  RRCrtcSet(pScrPriv->crtcs[i], mymode, new_x, new_y, RR_Rotate_0,  
> 1, &(pScrPriv->outputs[i]));
>
> this is the line I suggested to add, is this still required?

Yes. In fact, I didn't have that for a long time while testing and  
finally considered having it a good idea. Suddenly all remaining woes  
were gone.

> When I started all this I hade code cascades like this, too. But in  
> the end it all lead to basically the same behaviour. So I _feel_  
> that your change being too big. From my POV it is simply a matter of  
> the position not changing in case of a resize. That probably is the  
> only thing that needs to be reconsidered. But I may be totally  
> wrong, I have to dig deeper...

I indeed think, that the code can be simplified more. In fact, only  
width and height need to be adjusted, where the agent overlaps the  
bounding box. The rest should be handled fairly well via the  
intersect() function itself.

I need to boil this down more. Please test the code and see what my  
wanted end result is.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

sunweaver commented:
@uli42: One thought I had this morning, regarding this. If people have different monitor resolutions next to each other, then my sense is that the bounding box should not be an outer bounding box, but an inner bounding box. Inner bounding box: a box that can be placed inside the available screen, so that, if placed at the right location, all area of the inner bounding box is visible. Just a thought...

sunweaver commented:
@uli42: please take a second glance. I could simplify the patch tremendously and it works as wanted here.

sunweaver commented:
@uli42: I left the v1 of the patch in this PR for now to show the history of this PR. This PR should be rebased, of course, before merging.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3693,270 +3705,54 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
 
     /* check if outside-left of bounding box */
-    if (ax1 < bbx1) {
+    if (bx1 == bbx1 && ax1 - bbx1 < 0) {
```
are you sure that "-" takes precendence over "<"?  I'd expect this code too look like this:

if (bx1 == bbx1 && (ax1 - bbx1) < 0) {

uli42 replied:
Ok, looks like this is correct (see http://en.cppreference.com/w/c/language/operator_precedence). Still I'd like it to be enclosed...

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3693,270 +3705,54 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
-      fprintf(stderr, "intersect_bb: width stays unchanged [%d]\n", aw);
-      #endif
+     /* check if outside-right of bounding box */
+    if (bx1 + bw == bbx2 && ax1 + aw > bbx2) {
```
same here

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3693,270 +3705,54 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
 
     /* check if outside-above of bounding box */
-    if (ay1 < bby1) {
+    if (by1 == bby1 && ay1 - bby1 < 0) {
```
and here

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3693,270 +3705,54 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
-       * This mainly is: session window is y-wise on one CRTC only.
-       */
-      *h = ah;
+        *h = *h + (ay1 + ah - bby2);
```
what about `+=`?

Also using a variable for `(ay1 + ah - bby2)` might increase readbility, like this (lacking  a good variable name):

```
    int aystuff = ay1 + ah - bby2;
    int bystuff = by1 + bh - bby2;

    if (bystuff == 0 && aystuff > 0) {
        *h += aystuff;
```

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3693,270 +3705,54 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
-    if (ax1 < bbx1) {
+    if (bx1 == bbx1 && ax1 - bbx1 < 0) {
+
+        *w = *w + (bbx1 - ax1);
```
+=

sunweaver commented:
@uli42: like this? (next iteration...)

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Screen.c:
```diff
@@ -3718,9 +3718,9 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
 
      /* check if outside-right of bounding box */
-    if (bx1 + bw == bbx2 && ax1 + aw > bbx2) {
+    if (bx1 + bw == bbx2 && (ax1 + aw) > bbx2) {
```
would prefer parens for (bx1+bw), too. Not necessary, but looks more unified.

uli42 commented:
Still not happy with the code format. However, that is irrelevant for now. I need to think about the outcome of this change and do some tests. Cannot do that right now, may later today.

uli42 commented:
function-wise I am totally fine with this pr. But please add parens everywhere (or nowhere) in intersect_bb. Whatever you prefer.

sunweaver commented:
On  Mo 05 Mär 2018 15:46:42 CET, Ulrich Sibiller wrote:

> function-wise I am totally fine with this pr. But please add parens  
> everywhere (or nowhere) in intersect_bb. Whatever you prefer.

Not 100% sure where or where not you want more or less parentheses...

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

uli42 commented:
Either user `((a+b) < c && (d + e) > f)` or `(a+b < c && d+e >f)`. Currently we have a mix in intersect_bb. See also my comment above.

And while your are at it: please also fix the whitespace mess I seem to have introduced 2,5 years ago in Screen.c.

And one more improvement: the RRCrtcSet call that is added with this PR is now in both blocks of the if clause. It can - together with the preceding debug statement - easily be moved to after the if clause.

sunweaver commented:
@uli42: now ok? If so, I'll rebase and you merge(!). Ok?

uli42 commented:
The parens stuff is ok now. I have some more (more or less cosmetic) things to change in Xinerama: too many empty lines in intersect_bb, trailing whitespace, tab/space mix, RRSetCrtc line should be moved outside the if clause,...

Do you want to add this stuff or rather merge and let me make another PR?

sunweaver commented:
Hi Uli,

On  Di 06 Mär 2018 22:50:47 CET, Ulrich Sibiller wrote:

> The parens stuff is ok now. I have some more (more or less cosmetic)  
> things to change in Xinerama: too many empty lines in intersect_bb,  
> trailing whitespace, tab/space mix, RRSetCrtc line should be moved  
> outside the if clause. Do you want to add this stuff or rather merge  
> and let me make another PR?

I moved the RRCrtcSet function below the if clause. Branch has been  
rebased. Please go ahead and merged (please GPG-sign the merge, if  
help is needed, let me know).

For the white space changes, please file a separate PR. Thanks!

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

uli42 commented:
Yes, I need help in merging to master. Have done that only once IIRC.

Uli

Am 07.03.2018 06:46 schrieb "Mike Gabriel" <notifications@github.com>:

> Hi Uli,
>
> On Di 06 Mär 2018 22:50:47 CET, Ulrich Sibiller wrote:
>
> > The parens stuff is ok now. I have some more (more or less cosmetic)
> > things to change in Xinerama: too many empty lines in intersect_bb,
> > trailing whitespace, tab/space mix, RRSetCrtc line should be moved
> > outside the if clause. Do you want to add this stuff or rather merge
> > and let me make another PR?
>
> I moved the RRCrtcSet function below the if clause. Branch has been
> rebased. Please go ahead and merged (please GPG-sign the merge, if
> help is needed, let me know).
>
> For the white space changes, please file a separate PR. Thanks!
>
> Mike
> --
>
> DAS-NETZWERKTEAM
> mike gabriel, herweg 7, 24357 fleckeby
> mobile: +49 (1520) 1976 148
> landline: +49 (4354) 8390 139
>
> GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22 0782 9AF4 6B30 2577 1B31
> mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArcticaProject/nx-libs/pull/670#issuecomment-371029938>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGBw5MLXHlDzKh8dmHsqqERx5ZopQ85tks5tb3QbgaJpZM4SZ3d9>
> .
>

uli42 commented:
I have found a problem in the xinerama code, see https://imagebin.ca/v/3tzKxzPjhLZd

The light blue are in the lower left screen of the screenshot is invisible in reality on my setup because the left screen is only 1080 in height. That's all ok. The problem is, that our (fixed) xinerama code should increase the xinerama height of the left screen to the full 900 because is partly in the invisible area (I hope you understand the explanation).

I am not sure about the correct solution. Your suggestion about an inner bounding box could probably help here but will it in any case?

I _think_ we need to check these cases after the intersection
a) intersection edge is within the inner bounding box - nothing to do, take intersection result
b) intersection edge is outside the inner bounding box change intersection - increase height

(I have just tried and failed; turns out calculation of the inner bounding box is more complicated than it appears on first thought.)

sunweaver commented:
On  Mi 07 Mär 2018 20:35:33 CET, Ulrich Sibiller wrote:

> I have found a problem in the xinerama code, see  
> https://imagebin.ca/v/3tzKxzPjhLZd
>
> The light blue are in the lower left screen of the screenshot is  
> invisible in reality on my setup because the left screen is only  
> 1080 in height. That's all ok. The problem is, that our (fixed)  
> xinerama code should increase the xinerama height of the left screen  
> to the full 900 because is partly in the invisible area (I hope you  
> understand the explanation).
>
> I am not sure about the correct solution. Your suggestion about an  
> inner bounding box could probably help here but will it in any case?
>
> I _think_ we need to check these cases after the intersection
> a) intersection edge is within the inner bounding box - nothing to  
> do, take intersection result
> b) intersection edge is outside the inner bounding box change  
> intersection - increase height
>
> (I have just tried and failed; turns out calculation of the inner  
> bounding box is more complicated than it appears on first thought.)

please file this as a new bug...

And let's get #670 merged.

I'll respond with more thoughts on the above once I see the separate  
bug report on this.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de
